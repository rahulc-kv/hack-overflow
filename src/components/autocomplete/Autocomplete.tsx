import React, { FC, useEffect, useState } from 'react';
import { isEmpty } from 'lodash';
import { Controller } from 'react-hook-form';
import Autocomplete from '@mui/material/Autocomplete';
import Box from '@mui/material/Box';

import { TextField } from '@components';
import { DownArrowBlack } from '@assets/icons';
import { OptionsLOVType } from 'types/generalTypes';
import { AutoCompleteProps } from './types';

const AutocompleteDropdown: FC<AutoCompleteProps> = props => {
  const {
    name,
    placeholder,
    control,
    options,
    errors,
    selectedItem = null,
    onChangeHandler = () => null,
    isDisabled = false
  } = props;

  const [selectedValue, setSelectedValue] =
    useState<OptionsLOVType>(selectedItem);

  useEffect(() => {
    setSelectedValue(selectedItem);
  }, [selectedItem]);

  return (
    <div>
      <Controller
        name={name}
        control={control}
        render={({ field: { onChange } }) => (
          <Autocomplete
            sx={{
              width: '100%',
              '& .MuiAutocomplete-endAdornment': {
                top: 'calc(50% - 12px)',
                marginRight: '16px'
              },
              '& .MuiOutlinedInput-root': {
                padding: '8px'
              }
            }}
            options={options}
            autoHighlight={true}
            autoComplete={false}
            onChange={(event, value: OptionsLOVType) => {
              setSelectedValue(value);
              onChange(value?.id);
              onChangeHandler(value);
            }}
            popupIcon={<DownArrowBlack />}
            value={selectedValue}
            getOptionLabel={(option: OptionsLOVType) => {
              return option.name;
            }}
            isOptionEqualToValue={(option, value) => option?.id === value?.id}
            renderOption={(renderingprops, option) => (
              <Box
                component="li"
                sx={{ '& > img': { mr: 2, flexShrink: 0 } }}
                {...renderingprops}>
                {option.name}
              </Box>
            )}
            renderInput={params => (
              <TextField
                {...params}
                label={placeholder}
                name={name}
                errors={errors}
              />
            )}
            disabled={isDisabled}
          />
        )}
      />
      {!isEmpty(errors) && errors[name]?.message && (
        <p className="mt-[4px] text-xs text-red-500">{errors[name]?.message}</p>
      )}
    </div>
  );
};

export default AutocompleteDropdown;
