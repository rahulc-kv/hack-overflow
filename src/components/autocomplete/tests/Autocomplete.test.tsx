import React from 'react';
import { useForm } from 'react-hook-form';
import renderer from 'react-test-renderer';

import Autocomplete from '../Autocomplete';

test('Autocomplete Rendered', () => {
  const Component = () => {
    const { control } = useForm<{ value: string }>({
      defaultValues: {
        value: ''
      }
    });

    return (
      <Autocomplete name={''} placeholder={''} control={control} options={[]} />
    );
  };

  const component = renderer.create(<Component />).toJSON();

  expect(component).toMatchSnapshot();
});
