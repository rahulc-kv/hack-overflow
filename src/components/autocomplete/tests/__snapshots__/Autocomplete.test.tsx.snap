// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete Rendered 1`] = `
<div>
  <div
    aria-owns={null}
    className="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1jtatcx-MuiAutocomplete-root"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
  >
    <div
      className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root w-full css-bz5xn5-MuiFormControl-root-MuiTextField-root"
    >
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
        data-testid="text-field"
        onClick={[Function]}
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-expanded={false}
          aria-invalid={false}
          autoCapitalize="none"
          autoComplete="off"
          autoFocus={false}
          className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          name=""
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          readOnly={false}
          required={false}
          role="combobox"
          spellCheck="false"
          type="text"
          value=""
        />
        <div
          className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Open"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            title="Open"
            type="button"
          >
            <svg>
              downArrowBlack.svg
            </svg>
          </button>
        </div>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-hdw1oc"
          >
            <span
              className="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
